src/gg.regression.unity.bots/Runtime/Scripts/StateRecorder/BotSegments/BotSegmentsPlaybackController.cs

```diff
@@ -56,8 +58,351 @@ public class BotSegmentsPlaybackController : MonoBehaviour
 
         private ScreenRecorder _screenRecorder;
 
+        private ActionExplorationDriver _explorationDriver;
+
+        private void ProcessBotSegments()
+        {
+            var now = Time.unscaledTime;
+
+            try
+            {
+                if (_unpaused)
+                {
+                    _lastTimeLoggedKeyFrameConditions = now;
+                    _unpaused = false;
+                }
+
+                var objectFinders = FindObjectsByType<ObjectFinder>(FindObjectsSortMode.None);
+
+                Dictionary<long, ObjectStatus> transformStatuses = null;
+                Dictionary<long, ObjectStatus> entityStatuses = null;
+
+                foreach (var objectFinder in objectFinders)
+                {
+                    if (objectFinder is TransformObjectFinder)
+                    {
+                        transformStatuses = objectFinder.GetObjectStatusForCurrentFrame().Item2;
+                    }
+                    else
+                    {
+                        entityStatuses = objectFinder.GetObjectStatusForCurrentFrame().Item2;
+                    }
+                }
+
+                transformStatuses ??= new Dictionary<long, ObjectStatus>();
+                entityStatuses ??= new Dictionary<long, ObjectStatus>();
+
+                // track if any segment matched this update
+                var matchedThisUpdate = false;
+
+                // track if we have a new segment to evaluate... so long as we do, keep looping here before releasing from this Update call
+                // thus we process each new segment as soon as possible and don't have any artificial one frame delays before processing
+                var nextBotSegmentIndex = 0;
+                while (nextBotSegmentIndex < _nextBotSegments.Count)
+                {
+                    var nextBotSegment = _nextBotSegments[nextBotSegmentIndex];
+
+                    // if we're working on the first entry in the list is the only time we do actions
+                    if (nextBotSegmentIndex == 0)
+                    {
+                        ProcessBotSegmentAction(nextBotSegment, transformStatuses, entityStatuses);
```

Comment:
This line is the only real change of note in this method... the rest of it shows changed because I got tired of trying to find it in this class and moved it to the top